### ISSUE TITLE
Module naming declarations for %included interface files ignored?

### ISSUE THREAD

**Issue by Kagetsuki** [swig/swig#451]
I'm building a project with multiple modules which I'm naming based on their namespaces/classes as noted [here](http://www.swig.org/Doc3.0/Ruby.html#Ruby_nn12). I have the primary interface file that looks like this:

```
%module Root

%include "ns1.i"
%include "ns2.i"

namespace Root {
}
```

inside ns1.i and ns2.i are more %includes, for example an interface in Ns1 could be "Data" and that interface file would look like:

```
%module "Root::Ns1::Data"
// ^This can be anything, this %module directive seems to be ignored

namespace Root {
  namespace Ns1 {
    class Data
    {
      // Stuff...
    };
  }
}
```

*Keep in mind this is just a very basic example, my actual code isn't so primitive.

The generated wrapper however will not have Root::Ns1::Data, it will have Root::Data. This is regardless of what is defined for the module directive in ns1/data.i.

Any hints as to how I may remedy this? I'd like to have a clean structure of nested namespaces/modules generated for my wrappers so I can maintain the same conventions as my C++ base code in at least Ruby and Java. I can point to my actual code if needed. Any help would be greatly appreciated.

Side notes:
I'm using cmake and am only setting one SWIG_FLAGS for "-includeall".

**Comment by Kagetsuki** [comment]
Sorry to bump this but I really need to get this done. Basically I have a master namespace, two sub-namespaces, and classes nested under those namespaces. I'd like to be able to have:
Master::Sub1::ClassA
Master::Sub2::ClassB
but instead all I'm getting is:
Master::ClassA
Master::ClassB

The documentation says it's possible to generate nested namespaces/modules but I'm not able to do this - perhaps because I'm using %include directives to bundle all my interfaces? Any hints would be appreciated.

**Comment by ojwb** [comment]
`%include` essentially just performs a C-preprocessor like textual inclusion of the other file, so the situation is as if you'd specified `%module` multiple times in the same file.  If you do that, the name from the first is used from looking at the parser source code:

```
                 if (!ModuleName) ModuleName = NewString($3);
```

Though it looks like flags from them are merged if I follow the rest of the code.

Anyway, that's not going to do what you're hoping.

The Ruby docs say (http://swig.org/Doc3.0/Ruby.html#Ruby_nn22):

> SWIG is aware of C++ namespaces, but namespace names do not appear in the module nor do namespaces result in a module that is broken up into submodules or packages.

So this all seems to be working as documented.

**Comment by Kagetsuki** [comment]
@ojwb Aaaah, now that makes sense. I guess I'll try to compile each module separately then. Thanks!

**Comment by wsfulton** [comment]
Maybe you want the 'package' option in %module, see http://swig.org/Doc3.0/Python.html#Python_modulepackage. I think this only works for script languages. For Java use the nspace feature to map C++ namespaces to Java packages, see http://swig.org/Doc3.0/Java.html#Java_namespaces.

The nspace feature is only implemented in a few languages (Java, C#, D, Lua) maybe more, but not Ruby. It does need implementing. If you are interested let me know, it shouldn't be very difficult.

**Comment by Kagetsuki** [comment]
@wsfulton Thanks! I've been investigating this a bit and I'm not sure the 'package' option does anything in Ruby. I get the feeling this kind of functionality was implemented by including namespaces before the module name in quotes like:
```
%module "A::B::C"
```
which shows up in the documentation for Ruby. Perhaps this functionality should have been consolidated into one standard form if the semantics are the same - though I have a feeling they are not as package seems to imply nested named directories that correspond to the package naming structure (not unlike Java).

nspace on the other hand has an extremely averse side effect in Ruby! If you take the sample code:
```
%feature("nspace") MyWorld::Material::Color;
%nspace MyWorld::Wrapping::Color; // %nspace is a macro for %feature("nspace")

namespace MyWorld {
  namespace Material {
    class Color {
      int test(int num);
    };
  }
  namespace Wrapping {
    class Color {
      int test(int num);
    };
  }
}
```
And run that to ruby you will get:
"MyWorld::Color" which will contain a method named "MyWorld_Wrapping_Color_test". I seriously doubt this is what anyone would want - this should probably be remedied.

I'm going to test a sort of "multi module build" where I don't use cascading includes and see how that works. I think this should be possible but cmake doesn't seem to want to play nice and I'm having trouble finding an example of a project using such a build so I'm feeling my way through it.

As for implementing nspace in Ruby it would be for nothing if it didn't work on cascading includes. I'll try and test if it does work on Java and if it does I'll give you a heads up for some instructions on where to get started implementing for Ruby.

**Comment by wsfulton** [comment]
I don't really know if the module attribute will work for Ruby. Using colons in a module name seems wrong to me, but it is documented as such. I also don't see any tests for this except for director_nestedmodule.i which is only tested for Perl. Bottom line, is I really don't know if this is going to work and as I don't know Ruby very well, I can't experiment either. 

Regarding the package attribute in %module, there is a very comprehensive test in Examples/python/relativeimport and a runtime Perl test : packageoption_runme.pl for the packageoption test. Perhaps these might help you try the equivalent in Ruby.

As nspace isn't implemented in Ruby, the output is going to be completely wrong.

**Comment by Kagetsuki** [comment]
@wsfulton Using colons in the module name definitely works in Ruby whereas 'package' does not. This should probably all be consolidated into one feature, as in 'package' should probably work for Ruby as well. Otherwise each language will require completely different .i files which seems to defeat the purpose. Though on that note, perhaps nspace should be grouped into 'package' as well?

**Comment by wsfulton** [comment]
I can't see any code for Ruby that uses package in %module. Just Python and Perl use this and it looks like their use is subtly different. It looks like the -prefix option is the equivalent to Python and Perl's use of package in %module. See http://swig.org/Doc3.0/Ruby.html#Ruby_nn12. 

%module(package="...") was introduced way before %nspace. %nspace is designed to automatically map C++ namespaces into a target language equivalent for nested scoping. %nspace provides additional scoping for classes in C++ namespaces, but they will still be in the module name given by %module. For example:
```swig
%module mymodule
class C {};
namespace S { class K {}; }
```
will be accessible from target language as mymodule.C and mymodule.S.K (mymodule::C and mymodule::S::K in Ruby?). %nspace does work for Java, I implemented it for Java and C# and the implementation is missing in the other languages. I think efforts should focus on implementing %nspace for Ruby so you can get the automatic nesting you want. I can offer some sort of guidance on the SWIG side, but I don't really know Ruby well enough to know what the generated code should look like.

**Comment by Kagetsuki** [comment]
@wsfulton If you can give me a heads up as to where to start I guess I can give it a shot. I'm not very familiar with the actual inner-workings of SWIG though so I can't promise anything.

And for the example, currently if you add an %nspace it prefixes the *methods* within a class. So say you did '%nspace S::K' in the above example and you had method M in K you'd get:
mymodule::K::S_K_M()
Which is obviously NOT what anyone would want.

**Comment by wsfulton** [comment]
S_K_M type mangling will be needed in the C layer wrappers (the functions beginning with _wrap_). So it needs to be _wrap_S_K_M and not _wrap_K_M as it is now and should remain if %nspace is not used.

The work will all be in ruby.cxx. The NSPACE_TODO macros will need replacing with getNSpace(). After that, testing and modifying any more code that will get the scoping correct. Use the -debug-lsymbols option to view the Ruby symbol names. This will need to show the correct scoping too. Use java.cxx or csharp.cxx for a reference implementation, although I suspect this implementation was more involved than for Ruby. Do not use use lua.cxx as a reference implementation, but javascript.cxx might be helpful.

There are a number of nspace tests. Look for this:
```
// nspace feature only supported by these languages
#if defined(SWIGJAVA) || defined(SWIGCSHARP) || defined(SWIGD) || defined(SWIGLUA) || defined(SWIGJAVASCRIPT)
```
in the Examples/test-suite/*.i files and add in Ruby and provide runtime tests.

**Comment by Kagetsuki** [comment]
@wsfulton Awesome! I'll bang my head against this over the weekend and see what I can do.

**Comment by Kagetsuki** [comment]
I'm sorry to say I've spent quite a bit of my free time on this and for the life of me I can't figure it out. On top of that we're having terrible issues with using SWIG to get interfaces on Android. If there's someone out there familiar with SWIG who could give us a hand with the Android stuff I'll see what I can do with coming up with the funds to also get this implemented. Otherwise I'm afraid I've got no choice but to give up and just do multiple native implementations.

Thanks for all the help @wsfulton. Sorry I couldn't get this rolling.
